### PR TITLE
sequencer: refactor functions that manipulate the todo-list

### PR DESCRIPTION
**Pull request by agrn**
This refactors most functions that manipulate the todo-list not to read it from the filesystem. The relevant commits are the last nine.

I’m not sure that the new functions have a good name…

@chriscool @stefanbeller @dscho

### CODE REVIEW

File: sequencer.c [hunk 1]
```diff
@@ -4314,111 +4314,26 @@ int transform_todos(unsigned flags)
...
-}
-
-define_commit_slab(commit_seen, unsigned char);
-/*
- * Check if the user dropped some commits by mistake
- * Behaviour determined by rebase.missingCommitsCheck.
- * Check if there is an unrecognized command or a
- * bad SHA-1 in a command.
- */
 int check_todo_list(void)
```
Comment (chriscool) [inline sequencer.c]: Maybe renaming check_todo_list() to check_todo_list_from file() would make things clearer.

File: sequencer.c [hunk 2]
```diff
@@ -4314,111 +4314,26 @@ int transform_todos(unsigned flags)
...
 {
-	enum missing_commit_check_level check_level = get_missing_commit_check_level();
-	struct strbuf todo_file = STRBUF_INIT;
-	struct todo_list todo_list = TODO_LIST_INIT;
-	struct strbuf missing = STRBUF_INIT;
-	int advise_to_edit_todo = 0, res = 0, i;
-	struct commit_seen commit_seen;
-
-	init_commit_seen(&commit_seen);
+	struct todo_list old = TODO_LIST_INIT, new = TODO_LIST_INIT;
```
Comment (chriscool) [inline sequencer.c]: Nit: I would call the variable `old_todo` and `new_todo` instead of just `old` and `new`.

File: sequencer.c [hunk 3]
```diff
@@ -4710,18 +4700,42 @@ int rearrange_squash(void)
...
 
 	clear_commit_todo_item(&commit_todo);
+
+	if (todo_list_parse_insn_buffer(todo_list->buf.buf, todo_list))
+		BUG("unusable todo list");
+
+	return 0;
+}
+
+int rearrange_squash(void)
```
Comment (chriscool) [inline sequencer.c]: Maybe this could be called rearrange_squash_in_todo_file() or something like that.

**Comment by chriscool** [comment]
There are a few nits I noticed especially about naming functions that operate on files. Otherwise it looks good to me.

File: sequencer.c [hunk 4]
```diff
@@ -4324,7 +4314,31 @@ int transform_todos(unsigned flags)
...
+}
+
+int transform_todos(unsigned flags)
+{
+	const char *todo_file = rebase_path_todo();
+	struct todo_list todo_list = TODO_LIST_INIT;
+	int i;
+
+	if (strbuf_read_file(&todo_list.buf, todo_file, 0) < 0)
+		return error(_("could not read '%s'."), todo_file);
```
[the remark this answers is not in the document]
Comment (agrn) [reply sequencer.c]: I don’t understand why you think this is not related to the current patch, this code was part of the original ``transform_todos()``.

[the remark this answers is not in the document]
Comment (chriscool) [reply sequencer.c]: I mean that the current patch doesn't introduce the error) call, so it is a separate issue (from the issue the current patch addresses) to change error() into error_errno(), which means that changing error() into error_errno() should be done in another patch.

**Comment by agrn** [comment]
I updated my branch

**Comment by dscho** [comment]
For the record: the CI tests failed yesterday because of that [flaky t5570-osx problem that is already fixed in `pu`](https://github.com/git/git/compare/master...gitgitgadget:jk/no-sigpipe-during-network-transport). I re-ran the CI test and it passed (as flaky tests usually do when run again).

**Comment by dscho** [comment]
@agrn this has been integrated via e62e68d35949, right?

**Comment by agrn** [comment]
Yes.  I’m closing this.

**Pull request closed without merging**